state/protocol/inmem/snapshot.go

```diff
@@ -49,6 +49,10 @@ func (s Snapshot) SealedResult() (*flow.ExecutionResult, *flow.Seal, error) {
 	return s.enc.LatestResult, s.enc.LatestSeal, nil
 }
 
+func (s Snapshot) Seal() (*flow.Seal, error) {
+	return s.enc.SealingSegment.Seal, nil
```

Comment:
I'm not quite sure about this. 

What is the SealingSegment for a snapshot? Is the SealingSegment's first block equal to the snapshot's block? @jordanschalm 

If yes, then I'm right here, that if this snapshot is sealed, then the sealingsegment 's seal is the seal for the snapshot's block

Reply:
I also have mixed feeling about this. Originally, the Snapshot was supposed to be immutable. If retrieving a `Snapshot` instance is successful, then it should never change. The problem with the return value for `Seal()` is that it is eventually consistent. I am hoping that we could use https://github.com/onflow/flow-go/blob/b69fa33a25ed981e2ba77a33d3329cf0bb4e6fc7/storage/seals.go#L10-L24 instead

Reply:
>In general, I would prefer we avoid adding these kinds of methods to the Snapshot API 

Yeah, but the `Seal` method is just the same as the `QuorumCertificate` method, which is some data in future blocks.

Based on what Jordan mentioned, I think what I implemented is wrong. The `Seal` method should return the seal for the snapshot's reference block, but what I implemented is the last seal as of the snapshot's reference block.

```
A <- B <- C <- D (seal_A) <- E <- F <- G (seal_D)
```

If block D is the snapshot's reference block, then we should return `seal_D`, but my implementation is returning `seal_A`. 

As Alex suggested, we could choose to use seals storage interface if we don't like protocol's API and don't have a good way to add it.

Reply:
@zhangchiqing I think we should use the seals storage interface, because the usecase we have that originally motivated this PR is the one where given a block, we want to find its seal.

Reply:
Adding the `Seal()` method to the snapshot interface doesn't help satisfy this usecase.

Reply:
> Yeah, but the Seal method is just the same as the QuorumCertificate method, which is some data in future blocks.

You're right. The difference to me is that a QC is a necessary part of the snapshot which must be present in order to bootstrap a protocol state instance, whereas the seal is not. 

Some further thoughts:
* I think it would be reasonable to restrict the API so that you can only access a complete snapshot for a block with a valid QC (we need to do this for non-consensus nodes anyway, because we don't know that a block payload is valid until we observe the QC).
  * that way, we can include `QuorumCertificate` in the `Snapshot` API and maintain the property that a snapshot is immutable over time
  * however, there would need to be a carve-out for the HotStuff committee module to be able to query the `Identities` for a block without a validated QC
* Restricting the API in the same way to only sealed blocks is quite a bit more restrictive